### DISCUSSION TITLE
Widows and orphans

### DISCUSSION THREAD

**Discussion by pkvil** [olsak/OpTeX#98] [Q&A] [answered]
Hello. I'm in the process of migrating from ConTeXt to OpTeX, and I've managed to add all functionality I need except for one thing: adjusting vsize of specific pages to move lines from one page to another. I.e. widow control. 
The reason I need this is because I mainly typeset novels where there is no vertical stretch between lines or paragraphs, and the recommended way to deal with orphans and widows (Bringhurst, Tschichold) is to either shorten or lengthen a verso/recto pair of pages both by one line, to maintain balance.
In ConTeXt the high level command is \adaptlayout[lines=+1] and LaTeX has \enlargethispage{\baselineskip}.
I'm not very good with plain TeX (yet), but my current thinking is that I need to hijack the output routine before shipout, check if and how much to adjust vsize, unvbox255, and after shipout adjust vsize back to normal?
I'm also thinking that I'd rather not mess too much with the standard OpTeX output routine, instead I'd like to find a way to \addto the right part of it, if possible. If not possible at least learn where the natural place would be for this code to be added?
My question is: is my thinking about this reasonable or am I way off in how to do this? 
Any hints or ideas would be very welcome, thanks.

**Comment by vlasakm** [comment]
Sorry I only know basic things about widows and orphans, so maybe the following is not as relevant as I think, but for what its worth:

1. There is the [`lua-widow-control`](https://www.ctan.org/pkg/lua-widow-control) by Max Chernoff that works with plain, LaTeX, ConTeXt and even has a special support for OpTeX. The documentation seems like a great resource and the package provides automatic solution, though with compromises which I am not sure you are willing to make? There have been also some discussions on the ConTeXt mailing list about it recently, maybe you will find it interesting, though I didn't follow that closely.
2. From `lua-widow-control`'s documentation I found [a TUGboat article by Paul Isambert](https://www.tug.org/TUGboat/tb31-1/tb97isambert.pdf) which also discusses the different approaches including, as number 4, the one of Bringhurst as you mention.

As this is not my area of expertise, I probably can't help you much further.

Slightly off-topic: @olsak [`lua-widow-control`](https://www.ctan.org/pkg/lua-widow-control) seems like a good candidate for mention at the [OpTeX homepage](http://petr.olsak.net/optex/).

**Reply by pkvil** [reply]
Hi, and thank you, your links are very relevant. I have tried lua-widow-control and it's probably the best automated solution based on \looseness. It's especially nice because it keeps the number of lines constant. 
Unfortunately, what I need is a non-automated method of not keeping the number of lines constant. The problem with the looseness method is that sometimes the paragraphs available can't be affected by \looseness, and then you still need a manual intervention if you want to avoid widows. It's a finishing touch that is very hard to automate.

**Comment by olsak** [comment, marked as answer]
lua-widow-contol looks very nice. I did its work with \looseness manually so far. I described my method in my book TeXbook naruby (page 258, only in Czech).
If you need to give exceptions of \vsize at given pages (i.e. only manual intervention), then I can think about it and may be I prepare a new OpTeX trick. But if you want do implement more automatic solution then it is complicated task and I am not ready to give a result in a sort time.

**Reply by pkvil** [reply]
Hi, and maybe first of all I should thank you for OpTeX itself!
For this I specifically would like manual intervention only, no automation. 
Originally I tried to extract the relevant code from the (automatic) example in TeX by Topic on page 238 (to make it non-automatic), but I ran into two problems, 1) I'm not clever enough to fully understand it, and 2) I couldn't figure out where to put it in the OpTeX output routine.
Unfortunately I don't speak Czech, but I will try to read the code parts and see what I can learn.
Thank you.

**Reply by olsak** [reply]
OK, I prepared OpTeX trick, see http://petr.olsak.net/optex/optex-tricks.html#widows. Can it be like that?

About Czech TeXbook naruby: my notice was meant for my Czech colleague. The mentioned text in TeXbook naruby gets obsolete when the package lua-widow-control exists.

**Reply by pkvil** [reply]
Yes! That does exactly what I needed. Thank you so much!
